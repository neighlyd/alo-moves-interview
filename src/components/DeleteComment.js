import React from 'react';
import Modal from 'react-modal';

const DeleteComment = (props) => {

  return (
    <Modal
      isOpen={!!props.modalOpen}
      className="modal"
      closeTimeoutMS={200}
      contentLabel={props.formTitle}
      onRequestClose={props.handleModalClose}
    >
    <div>
      <h3>Delete This Comment?</h3>
      <button onClick={props.handleSubmit} className="delete__confirm_btn">Yes</button>
      <button onClick={props.handleModalClose} className="delete__deny_btn">No</button>
    </div>
    </Modal>
      
  )
}

export default DeleteComment;