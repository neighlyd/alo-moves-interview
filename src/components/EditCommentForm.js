import React, { useState } from 'react';
import Modal from 'react-modal'

const CommentForm = (props) => {

  const [ comment, setComment ] = useState(props.comment);

  const origComment = props.comment;

  const handleSubmit = e => {
    e.preventDefault();
    if (comment !== '' && comment !== origComment) {
      props.handleSubmit({text: comment});
    }
  };

  return (
    <Modal
      isOpen={!!props.modalOpen}
      className="modal"
      closeTimeoutMS={200}
      contentLabel={props.formTitle}
      onRequestClose={props.handleModalClose}
    >
      <div>
        <h3>{props.formTitle}</h3>
        <form onSubmit={handleSubmit} className='edit-comment'>
          <label
            className='edit-comment__label'
          >
            Comment
          </label>
          <textarea 
            rows='4'
            className='edit-comment__input text-input'
            autoFocus
            value={comment}
            onChange={e => setComment(e.target.value)}
          />
          { (comment !== '' && comment !== props.comment) ? (
            <button 
              className='edit-comment__btn'
            >
              {props.buttonLabel}
            </button> 
          ) : (
            <button 
              disabled
              className='edit-comment__btn'
            >
              {props.buttonLabel}
            </button>
          )}
        </form>
      </div>
    </Modal>
  )
}

export default CommentForm;