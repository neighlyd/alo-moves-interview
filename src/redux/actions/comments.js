import firebase from '../../firebase/firebase';
const db = firebase.collection('comments')

// Use const variables for action names, so we have a single place to edit them in the future.
export const ADD_COMMENT = 'ADD_COMMENT';
export const DELETE_COMMENT = 'DELETE_COMMENT';
export const EDIT_COMMENT = 'EDIT_COMMENT';
export const SET_COMMENTS = 'SET_COMMENTS';

// These are the pure functions that will be dispatched to the Redux store itself.
export const addComment = (comment) => ({
  type: ADD_COMMENT,
  comment
});

export const deleteComment = (id) => ({
  type: DELETE_COMMENT,
  id
});

export const editComment = (id, updates) => ({
  type: EDIT_COMMENT,
  id,
  updates
});

export const setComments = (comments) => ({
  type: SET_COMMENTS,
  comments
})

export const startAddComment = (comment = '') => {
  return async dispatch => {  
    try {      
      const data = {
        text: comment,
        updatedAt: new Date()
      };
      // Firestore's .add() method returns a documentReference. We need to populate this reference with the newly created object using the .get() method.
      // Once we have a populated object, we can then dispatch it to our Redux store.
      const docRef = await db.add(data);
      const doc = await docRef.get();
      
      if (doc) {
        dispatch(addComment({
          id: doc.id,
          text: doc.data().text,
          updatedAt: doc.data().updatedAt
        }));
      }
    } catch (err){
      console.log(err);
    };
  };
};

export const startDeleteComment = id => {
  return async dispatch => {  
    try{
      
      // First try to retrieve the Document from the Firestore DB.
      // If it still exists there, delete it remotely and locally.
      // If it has already been removed from remote, then only remove from local (e.g., another user deleted the comment and the page hasn't refreshed yet)
      // The need for this check could be partially alleviated by using sockets or subscribing to the DB store itself.
      const docRef = await db.doc(id).get();
      
      if (docRef.exists) {
        await db.doc(id).delete();
      } 
      dispatch(deleteComment(id));
    } catch (err) {
      console.log(err);
    };
  };
};

export const startEditComment = (id, comment) => {
  return async dispatch => {
    try {
      // As above, we error check to see if the document has been deleted by another user.
      const docRef = await db.doc(id).get();
      if (!docRef.exists) {
        // If another user deleted the comment before our user refreshes the page and attempts to edit, then their local version should be removed as well.
        // Otherwise, our local store will be out of sync.
        dispatch(deleteComment(id));
      } else {
        const updates = {
          ...comment,
          updatedAt: new Date()
        }
        await db.doc(id).set(updates)

        // Firestore does not return a document upon set(), so we need to retrieve the updated document from the remote DB.
        // We need to do this because they auto-format timestamp fields in their own format, which is incompatible with javascript's native Date() type.
        const doc = await db.doc(id).get();
        
        if (doc.exists) {
          const updates = {
            text: doc.data().text,
            updatedAt: doc.data().updatedAt
          };
        
          dispatch(editComment(doc.id, updates));
        } 
      }
    } catch (err) {
      console.log(err);
    };
  };
};

export const startSetComments = () => {
  return async dispatch => {
    try {
      
      // Start by fetching all comments from Firestore.
      // we then need to parse the data into an array (Firestore doesn't store info in arrays - at least, it's a pain to do so)
      // Finally, we dispatch() SET_EXPENSES with the new data array.
      let comments = [];
      
      const docRef = await db.get()
      
      docRef.forEach(comment => {
        comments.push({
          id: comment.id,
          text: comment.data().text,
          updatedAt: comment.data().updatedAt
        });
      });
      
      dispatch(setComments(comments));
    
    } catch (err) {
      console.log(err);
    };
  };
};