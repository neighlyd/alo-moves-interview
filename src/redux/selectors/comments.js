export default (comments) => {
  return comments.sort((a,b) => {
    return a.updatedAt < b.updatedAt ? -1 : 1;
  });
};