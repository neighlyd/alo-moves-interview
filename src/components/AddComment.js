import React, { useState } from 'react';
import { connect } from 'react-redux';

import { startAddComment } from '../redux/actions/comments';

export const AddComment = (props) => {
  const [ comment, setComment ] = useState('')

  const handleSubmit = (e) => {
    e.preventDefault();
    props.startAddComment(comment);
    setComment('')
  }

  return(
    <form onSubmit={handleSubmit} className='add-comment__form'>
      <input 
        type='text'
        placeholder='Enter Your Comment'
        className='add-comment__input text-input'
        aria-label='Add Comment'
        autoFocus
        value={comment}
        onChange={e => setComment(e.target.value)}
      />
      { comment !== '' ? (
        <button
          className='add-comment__btn'
        >
          Add Comment
        </button> 
      ) : (
        <button 
          disabled
          className='add-comment__btn'
        >
          Add Comment
        </button>
      )}
    </form>
  )
};

const mapDispatchToProps = (dispatch) => ({
  startAddComment: (comment) => dispatch(startAddComment(comment))
})

export default connect(null, mapDispatchToProps)(AddComment);
