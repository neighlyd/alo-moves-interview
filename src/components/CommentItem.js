import React, { useState } from 'react';
import { connect } from 'react-redux';
import * as linkify from 'linkifyjs';
import hashtag from 'linkifyjs/plugins/hashtag';
import Linkify from 'linkifyjs/react';

import EditCommentForm from './EditCommentForm';
import DeleteComment from './DeleteComment';
import { startDeleteComment, startEditComment } from '../redux/actions/comments';
import linkifyOpts from '../linkifyOpts';


export const CommentItem = (props) => {
  
  // One of the downsides to using modals like this is that we are littering our space. 
  // That is to say, every list item will create a <div class='ReactModalPortal'></div> element at the foot of our page. 
  // These elements will only be populated upon modal open, but it still has the potential to be a LOT of <div> elements.
  const [ editModalOpen, setEditModalOpen ] = useState(null);
  
  const handleEditModalClose = () => setEditModalOpen(false);
  
  const handleEditComment = (comment) => {
    props.startEditComment(props.id, comment);
    handleEditModalClose();
  }
  
  const editComment = () => {
    setEditModalOpen(true);
  }
  
  const [ deleteModalOpen, setDeleteModalOpen ] = useState(null);
  
  const handleDeleteModalClose = () => setDeleteModalOpen(false);
  
  const handleDeleteComment = () => {
    props.startDeleteComment(props.id);
    handleDeleteModalClose();
  }
  
  const deleteComment = () => {
    setDeleteModalOpen(true);
  }
  
  hashtag(linkify);
  
  return(
    <li>
      <div className='comment'>
        <Linkify options={linkifyOpts} className='comment__text'>{props.text}</Linkify>
        <div className='comment__buttons'>
          <button onClick={editComment}>Edit</button>
          <button onClick={deleteComment}>Delete</button>
        </div>
      </div>
      <EditCommentForm
        modalOpen={editModalOpen}
        handleModalClose={handleEditModalClose}
        formTitle='Edit Comment'
        buttonLabel='Save'
        comment={props.text}
        handleSubmit={handleEditComment}/>
      <DeleteComment
        modalOpen={deleteModalOpen}
        handleModalClose={handleDeleteModalClose}
        formTitle='Delete Comment'
        handleSubmit={handleDeleteComment}/>
    </li>
  )
};

const mapDispatchToProps = (dispatch) => ({
  startEditComment: (id, comment) => dispatch(startEditComment(id, comment)),
  startDeleteComment: (id) => dispatch(startDeleteComment(id))
})

export default connect(null, mapDispatchToProps)(CommentItem);