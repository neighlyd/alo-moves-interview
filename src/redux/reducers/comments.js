import { ADD_COMMENT, DELETE_COMMENT, EDIT_COMMENT, SET_COMMENTS } from '../actions/comments';

// Comments will be an array of objects. As such:
// let comments = [{id: 1, text: 'comment 1'}, {id: 2, text: 'comment 2'}]

const commentsReducerDefaultState = [];

export default (state = commentsReducerDefaultState, action) => {
  switch(action.type){
    case ADD_COMMENT:
      return [
        ...state,
        action.comment
      ];
    case DELETE_COMMENT:
      return state.filter( comment => comment.id !== action.id);
    case EDIT_COMMENT:
      return state.map(comment => {
        if (comment.id === action.id){
          return {
            ...comment,
            ...action.updates
          }
        } else {
          return comment;
        }
      });
    case SET_COMMENTS:
      return action.comments
    default:
      return state;
  };
};