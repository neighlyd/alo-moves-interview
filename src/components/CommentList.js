import React from 'react';
import { connect } from 'react-redux';

import CommentItem from './CommentItem';
import AddComment from './AddComment';
import commentsSort from '../redux/selectors/comments';

export const CommentList = (props) => {

  return(
    <div>
      <h2 className='comment-header'>Comments</h2>      
      <ul>
        { props.comments.length === 0 ? (
          <li>
            <p className='no-comment'>There are no comments yet. You can be the first commenter!</p>
          </li>
        ) : (
          props.comments.map(comment => (
            <CommentItem
              key={comment.id}
              {...comment}
              />
          ))
        )}
      </ul>
      <AddComment />
    </div>
  )
};

const mapStateToProps = (state) => {
  return {
    comments: commentsSort(state.comments)
  }
};

export default connect(mapStateToProps)(CommentList);