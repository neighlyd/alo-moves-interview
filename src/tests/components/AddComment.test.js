import React from 'react';
import { shallow } from 'enzyme';

import { AddComment } from '../../components/AddComment';
import comments from '../fixtures/comments';

let startAddComment, wrapper;

beforeEach(() => {
  startAddComment = jest.fn();
  wrapper = shallow(<AddComment startAddComment={startAddComment}/>);
});

it('AddComment should render correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

it('AddComment should handlSubmit correctly', () => {
  const newComment = 'New Comment';
  const input = wrapper.find('.add-comment__input');
  input.simulate('change', {target: {value: newComment}});

  wrapper.find('.add-comment__form').simulate('submit', {preventDefault: () => null});
  expect(startAddComment).toHaveBeenCalledWith(newComment);
})