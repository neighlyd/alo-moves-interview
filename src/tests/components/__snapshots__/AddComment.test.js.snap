// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddComment should render correctly 1`] = `ShallowWrapper {}`;
